README
NAME
    Catalyst::Engine::Apache - Catalyst Apache Engines

SYNOPSIS
    For example Apache configurations, see the documentation for the engine
    that corresponds to your Apache version.

    "Catalyst::Engine::Apache::MP13" - mod_perl 1.3x

    "Catalyst::Engine::Apache2::MP19" - mod_perl 1.99x

    "Catalyst::Engine::Apache2::MP20" - mod_perl 2.x

DESCRIPTION
    These classes provide mod_perl support for Catalyst.

METHODS
  $c->engine->apache
    Returns an "Apache", "Apache::RequestRec" or "Apache2::RequestRec"
    object, depending on your mod_perl version. This method is also
    available as $c->apache.

  $c->engine->return
    If you need to return something other than OK from the mod_perl handler,
    you may set any other Apache constant in this method. You should only
    use this method if you know what you are doing or bad things may happen!
    For example, to return DECLINED in mod_perl 2:

        use Apache2::Const -compile => qw(DECLINED);
        $c->engine->return( Apache2::Const::DECLINED );

  NOTES ABOUT LOCATIONMATCH
    The Apache engine tries to figure out the correct base path if your app
    is running within a LocationMatch block. For example:

        <LocationMatch ^/match/(this|that)*>
            SetHandler          modperl
            PerlResponseHandler MyApp
        </LocationMatch>

    This will correctly set the base path to '/match/this/' or
    '/match/that/' depending on which path was used for the request.

    In some cases this may not be what you want, so you can disable this
    behavior by adding this to your configuration:

        PerlSetVar CatalystDisableLocationMatch 1

  NOTES ON NON-STANDARD PORTS
    If you wish to run your site on a non-standard port you will need to use
    the "Port" Apache config rather than "Listen". This will result in the
    correct port being added to urls created using "uri_for".

        Port 8080

OVERLOADED METHODS
    This class overloads some methods from "Catalyst::Engine".

    prepare_request($r)
    prepare_connection
    prepare_query_parameters
    prepare_headers
    prepare_path
    read_chunk
    finalize_body
    finalize_headers
    write

SEE ALSO
    Catalyst Catalyst::Engine.

MAINTAINERS
    Current maintainer, Tomas Doran (t0m) "<bobtfish@bobtfish.net>".

AUTHORS
    Sebastian Riedel, "<sri@cpan.org>".

    Christian Hansen, "<ch@ngmedia.com>".

    Andy Grundman, "<andy@hybridized.org>".

COPYRIGHT
    Copyright 2005-2010 by the above above listed "AUTHORS" and
    "MAINTAINERS".

LICENSE
    This program is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
This file documents the revision history for Catalyst::Engine::Apache.

1.15   2010-10-03 22:07:00
        - Fix license info for debian et all.

1.14
        - No changes since dev release.

1.13_01
        - Update tests to pass again.
        - Update tests to not throw warnings with newer versions of Catalyst.
        - Fix $c->req->remote_user with this engine.
        - Clarify use with non-standard ports.
        - Fixes RT#61707, RT#61706, RT#61704, RT#36829

1.13
        - Workaround change in LWP that broke a cookie test (RT #40037)
        - Update streaming test to latest 5.70

1.12    2008-02-20 09:00:00
        - Fixed bug where %2b in query parameter is doubly decoded to ' ', instead of '+' 
          (Gavin Henry, Tatsuhiko Miyagawa)
        - Fixed bug where req->base and req->uri would include a port number when running 
          in SSL mode.
        - Fixed warning when a non-trivial regex is used in a LocationMatch block.
          (Steffen Schwigon)
        - Clear the value of $c->engine->return() before each request.
        - Added support for 'PerlSetVar CatalystDisableLocationMatch 1' to disable check
          for a LocationMatch regex to set base path.

1.11    2007-05-18 08:30:00
        - Don't 'use mod_perl;' as this may not work on some mod_perl installations.

1.10    2007-05-15 17:40:00
        - Properly detect the base when running within a LocationMatch block.
        - Use the unparsed URI for building the path instead of Apache's
          pre-parsed URI.
        - Load APR::Table when using mod_perl 1.99.
        - Switch to Module::Install.

1.09    2007-03-28 23:00:00
        - Fixed compatibility with older Catalyst versions.  5.7008+
          is recommended for best performance.

1.08    2007-03-28 22:45:00
        - Improved prepare_query_parameters performance by using C-based
          Apache modules instead of URI and URI::Escape.
        - Improved prepare_path performance by removing the use of URI.pm.
        - Extract host and port from X-Forwarded-Host.
        - Use Apache2::ModSSL for determining SSL mode if available.

1.07    2006-02-17 17:00:00
        - Fixed bug: Can't locate object method "FIRSTKEY" via package
          "APR::Table" when running under mod_perl 2.0.2.

1.06    2006-01-17 16:30:00
        - Removed t/01use.t as it will fail if Catalyst is not installed.

1.05    2005-12-23 10:00:00
        - Fixed double-cookie bug on redirects.
        - Synced tests with Catalyst.

1.04    2005-12-22 13:15:00
        - Fixed bug when processing a bare HTTP/1.0 HEAD request with no
          headers.

1.03    2005-12-13 22:30:00
        - Removed dependency on Catalyst because it causes an endless loop
          when trying to install Cat with AutoInstall.

1.02    2005-12-06 08:35:00
        - Default to text/html, like the standalone server.

1.01    2005-12-05 14:25:00
        - Added $c->engine->return method to allow custom Apache status
          codes to be returned from the mod_perl handler.
        - Fixed mod_perl 1.99 engine to use correct Apache::OK status code.

1.00    2005-11-13 19:30:00
        - No changes, version bumped to coincide with Catalyst 5.50.

0.99002 2005-11-08 09:00:00
        - Really fixed the return values and status codes this time!

0.99001 2005-10-25 21:25:00
        - Fixed return value from handler to use the correct
          Apache::Const or Apache2::Const value rather than a numeric
          value.
        - Fixed query param handling.
        - Added automated tests using Apache::Test.

0.99    2005-10-10 10:15:00
        - Initial release, separated Apache engines from Catalyst core.


t/optional/mod_perl.pl
#!perl

# Run all tests against Apache mod_perl
#
# Note, to get this to run properly, you may need to give it the path to your
# httpd.conf:
# 
# perl t/optional/mod_perl.pl -httpd_conf /etc/apache/httpd.conf
#
# For debugging, you can start TestApp and leave it running with
# perl t/optional/mod_perl.pl -httpd_conf /etc/apache/httpd.conf --start-httpd
#
# To stop it:
# perl t/optional/mod_perl.pl -httpd_conf /etc/apache/httpd.conf --stop-httpd

use strict;
use warnings;

use Apache::Test;
use Apache::TestRunPerl ();

$ENV{CATALYST_SERVER} = 'http://localhost:8529';

Apache::TestRunPerl->new->run(@ARGV);


t/optional/mod_perl-non-root.pl
#!perl

# Run all tests against Apache mod_perl
#
# Note, to get this to run properly, you may need to give it the path to your
# httpd.conf:
# 
# perl t/optional/mod_perl-non-root.pl -httpd_conf /etc/apache/httpd.conf
#
# For debugging, you can start TestApp and leave it running with
# perl t/optional/mod_perl.pl -httpd_conf /etc/apache/httpd.conf --start-httpd
#
# To stop it:
# perl t/optional/mod_perl.pl -httpd_conf /etc/apache/httpd.conf --stop-httpd

use strict;
use warnings;

use Apache::Test;
use Apache::TestRunPerl ();

$ENV{CATALYST_SERVER} = 'http://localhost:8529/deep/path';

Apache::TestRunPerl->new->run(@ARGV);


t/optional/mod_perl-locationmatch.pl
#!perl

# Run all tests against Apache mod_perl using a LocationMatch regex block
#
# Note, to get this to run properly, you may need to give it the path to your
# httpd.conf:
# 
# perl t/optional/mod_perl-locationmatch.pl -httpd_conf /etc/apache/httpd.conf
#
# For debugging, you can start TestApp and leave it running with
# perl t/optional/mod_perl-locationmatch.pl -httpd_conf /etc/apache/httpd.conf --start-httpd
#
# To stop it:
# perl t/optional/mod_perl-locationmatch.pl -httpd_conf /etc/apache/httpd.conf --stop-httpd

use strict;
use warnings;

use Apache::Test;
use Apache::TestRunPerl ();

$ENV{CATALYST_SERVER} = 'http://localhost:8529/match/that';

Apache::TestRunPerl->new->run(@ARGV);


lib/Catalyst/Engine/Apache2.pm
package Catalyst::Engine::Apache2;

use strict;
use warnings;
use base 'Catalyst::Engine::Apache';

sub finalize_headers {
    my ( $self, $c ) = @_;

    $self->SUPER::finalize_headers( $c );

    # This handles the case where Apache2 will remove the Content-Length
    # header on a HEAD request.
    # http://perl.apache.org/docs/2.0/user/handlers/http.html
    if ( $self->apache->header_only ) {
        $self->apache->rflush;
    }

    return 0;
}

1;
__END__

=head1 NAME

Catalyst::Engine::Apache2 - Base class for Apache 1.99x and 2.x Engines

=head1 SYNOPSIS

See L<Catalyst>.

=head1 DESCRIPTION

This is a base class for Apache 1.99x and 2.x Engines.

=head1 OVERLOADED METHODS

This class overloads some methods from C<Catalyst::Engine>.

=over 4

=item finalize_headers

=back

=head1 SEE ALSO

L<Catalyst> L<Catalyst::Engine>.

=head1 AUTHORS

Sebastian Riedel, <sri@cpan.org>

Christian Hansen, <ch@ngmedia.com>

Andy Grundman, <andy@hybridized.org>

=head1 COPYRIGHT

This program is free software, you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/Catalyst/Engine/Apache/MP13.pm
package Catalyst::Engine::Apache::MP13;

use strict;
use warnings;
use base 'Catalyst::Engine::Apache';

use Apache            ();
use Apache::Constants qw(OK);
use Apache::File      ();
use Apache::Util      ();

sub finalize_headers {
    my ( $self, $c ) = @_;
    
    $self->SUPER::finalize_headers( $c );
    
    $self->apache->send_http_header;
    
    return 0;
}

sub ok_constant { Apache::Constants::OK }

sub unescape_uri {
    my $self = shift;

    # Unlike in mod_perl 2, this method also unescapes '+' to space
    return Apache::Util::unescape_uri_info(@_);
}

1;
__END__

=head1 NAME

Catalyst::Engine::Apache::MP13 - Catalyst Apache mod_perl 1.3x Engine

=head1 SYNOPSIS

    # Set up your Catalyst app as a mod_perl 1.3x application in httpd.conf
    <Perl>
        use lib qw( /var/www/MyApp/lib );
    </Perl>
    
    # Preload your entire application
    PerlModule MyApp
    
    <VirtualHost *>
        ServerName   myapp.hostname.com
        DocumentRoot /var/www/MyApp/root
        
        <Location />
            SetHandler       perl-script
            PerlHandler      MyApp
        </Location>
        
        # you can also run your app in any non-root location
        <Location /some/other/path>
            SetHandler      perl-script
            PerlHandler     MyApp
        </Location>
        
        # Make sure to let Apache handle your static files
        # (And remember to remove the Static::Simple plugin in production)
        <Location /static>
            SetHandler      default-handler
        </Location>
    </VirtualHost>
    
=head1 DESCRIPTION

This is the Catalyst engine specialized for Apache mod_perl version 1.3x.

=head1 Apache::Registry

While this method is not recommended, you can also run your Catalyst
application via an Apache::Registry script.

httpd.conf:

    PerlModule Apache::Registry
    Alias / /var/www/MyApp/script/myapp_registry.pl/
    
    <Directory /var/www/MyApp/script>
        Options +ExecCGI
    </Directory>
    
    <Location />
        SetHandler  perl-script
        PerlHandler Apache::Registry
    </Location>
    
script/myapp_registry.pl (you will need to create this):

    #!/usr/bin/perl
    
    use strict;
    use warnings;
    use MyApp;
    
    MyApp->handle_request( Apache->request );
    
=head1 METHODS

=head2 ok_constant

=head1 OVERLOADED METHODS

This class overloads some methods from C<Catalyst::Engine::Apache>.

=head2 $c->engine->finalize_headers

=head1 SEE ALSO

L<Catalyst>, L<Catalyst::Engine>, L<Catalyst::Engine::Apache>.

=head1 AUTHORS

Sebastian Riedel, <sri@cpan.org>

Christian Hansen, <ch@ngmedia.com>

Andy Grundman, <andy@hybridized.org>

=head1 COPYRIGHT

This program is free software, you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/Catalyst/Engine/Apache2/MP19.pm
package Catalyst::Engine::Apache2::MP19;

use strict;
use warnings;
use base 'Catalyst::Engine::Apache2';

use Apache2             ();
use Apache::Connection  ();
use Apache::Const       -compile => qw(OK);
use Apache::RequestIO   ();
use Apache::RequestRec  ();
use Apache::RequestUtil ();
use Apache::Response    ();
use Apache::URI         ();
use APR::Table          ();

sub ok_constant { Apache::OK }

sub unescape_uri {
    my ( $self, $str ) = @_;
    
    $str =~ s/\+/ /g;
    return Apache::URI::unescape_url($str);
}

1;
__END__

=head1 NAME

Catalyst::Engine::Apache2::MP19 - Catalyst Apache2 mod_perl 1.99x Engine

=head1 SYNOPSIS

    # Set up your Catalyst app as a mod_perl 1.99x application in httpd.conf
    PerlSwitches -I/var/www/MyApp/lib
    
    # Preload your entire application
    PerlModule MyApp
    
    <VirtualHost *>
        ServerName    myapp.hostname.com
        DocumentRoot  /var/www/MyApp/root
        
        <Location />
            SetHandler          modperl
            PerlResponseHandler MyApp
        </Location>
        
        # you can also run your app in any non-root location
        <Location /some/other/path>
            SetHandler          perl-script
            PerlResponseHandler MyApp
        </Location>
        
        # Make sure to let Apache handle your static files
        # (It is not necessary to remove the Static::Simple plugin 
        # in production; Apache will bypass Static::Simple if
        # configured in this way)

        <Location /static>
            SetHandler          default-handler
        </Location>
 
        # If not running at a root location in a VirtualHost,
        # you'll probably need to set an Alias to the location
        # of your static files, and allow access to this location:

        Alias /myapp/static /filesystem/path/to/MyApp/root/static
        <Directory /filesystem/path/to/MyApp/root/static>
            allow from all
        </Directory>
        <Location /myapp/static>
            SetHandler default-handler
        </Location>

    </VirtualHost>

=head1 DESCRIPTION

This is the Catalyst engine specialized for Apache2 mod_perl version 1.99x.

=head1 WARNING

mod_perl 1.99 was the beta version for mod_perl 2.0.  Upgrading to 2.0 is
strongly recommended.

=head1 CGI ISSUES

In order to run Catalyst under mod_perl 1.99 you will need to downgrade L<CGI>
to version 3.07, as it has dropped support for 1.99 in later versions.

=head1 ModPerl::Registry

While this method is not recommended, you can also run your Catalyst
application via a ModPerl::Registry script.

httpd.conf:

    PerlModule ModPerl::Registry
    Alias / /var/www/MyApp/script/myapp_registry.pl/
    
    <Directory /var/www/MyApp/script>
        Options +ExecCGI
    </Directory>
    
    <Location />
        SetHandler          perl-script
        PerlResponseHandler ModPerl::Registry
    </Location>
    
script/myapp_registry.pl (you will need to create this):

    #!/usr/bin/perl
    
    use strict;
    use warnings;
    use MyApp;
    
    MyApp->handle_request( Apache::RequestUtil->request );
    
=head1 METHODS

=head2 ok_constant

=head1 SEE ALSO

L<Catalyst>, L<Catalyst::Engine>, L<Catalyst::Engine::Apache2>.

=head1 AUTHORS

Sebastian Riedel, <sri@cpan.org>

Christian Hansen, <ch@ngmedia.com>

Andy Grundman, <andy@hybridized.org>

=head1 COPYRIGHT

This program is free software, you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/Catalyst/Engine/Apache2/MP20.pm
package Catalyst::Engine::Apache2::MP20;

use strict;
use warnings;
use base 'Catalyst::Engine::Apache2';

use Apache2::Connection  ();
use Apache2::Const       -compile => qw(OK);
use Apache2::RequestIO   ();
use Apache2::RequestRec  ();
use Apache2::RequestUtil ();
use Apache2::Response    ();
use Apache2::URI         ();
use APR::Table           ();

# We can use Apache2::ModSSL to better detect if we're running in SSL mode
eval { require Apache2::ModSSL };

sub ok_constant { Apache2::Const::OK }

sub unescape_uri {
    my ( $self, $str ) = @_;
    
    $str =~ s/\+/ /g;
    return Apache2::URI::unescape_url($str);
}

1;
__END__

=head1 NAME

Catalyst::Engine::Apache2::MP20 - Catalyst Apache2 mod_perl 2.x Engine

=head1 SYNOPSIS

    # Set up your Catalyst app as a mod_perl 2.x application in httpd.conf
    PerlSwitches -I/var/www/MyApp/lib
    
    # Preload your entire application
    PerlModule MyApp
    
    <VirtualHost *>
        ServerName    myapp.hostname.com
        DocumentRoot  /var/www/MyApp/root
        
        <Location />
            SetHandler          modperl
            PerlResponseHandler MyApp
        </Location>
        
        # you can also run your app in any non-root location
        <Location /some/other/path>
            SetHandler          perl-script
            PerlResponseHandler MyApp
        </Location>
        
        # Make sure to let Apache handle your static files
        # (It is not necessary to remove the Static::Simple plugin 
        # in production; Apache will bypass Static::Simple if
        # configured in this way)

        <Location /static>
            SetHandler          default-handler
        </Location>
 
        # If not running at a root location in a VirtualHost,
        # you'll probably need to set an Alias to the location
        # of your static files, and allow access to this location:

        Alias /myapp/static /filesystem/path/to/MyApp/root/static
        <Directory /filesystem/path/to/MyApp/root/static>
            allow from all
        </Directory>
        <Location /myapp/static>
            SetHandler default-handler
        </Location>

    </VirtualHost>

=head1 DESCRIPTION

This is the Catalyst engine specialized for Apache2 mod_perl version 2.x.

=head1 ModPerl::Registry

While this method is not recommended, you can also run your Catalyst
application via a ModPerl::Registry script.

httpd.conf:

    PerlModule ModPerl::Registry
    Alias / /var/www/MyApp/script/myapp_registry.pl/
    
    <Directory /var/www/MyApp/script>
        Options +ExecCGI
    </Directory>
    
    <Location />
        SetHandler          perl-script
        PerlResponseHandler ModPerl::Registry
    </Location>
    
script/myapp_registry.pl (you will need to create this):

    #!/usr/bin/perl
    
    use strict;
    use warnings;
    use MyApp;
    
    MyApp->handle_request( Apache2::RequestUtil->request );

=head1 METHODS

=head2 ok_constant

=head1 OVERLOADED METHODS

This class overloads some methods from C<Catalyst::Engine>.

=over 4

=item unescape_uri 

=back

=head1 OVERLOADED METHODS

This class overloads some methods from C<Catalyst::Engine>.

=over 4

=item unescape_uri 

=back

=head1 SEE ALSO

L<Catalyst>, L<Catalyst::Engine>, L<Catalyst::Engine::Apache2>.

=head1 AUTHORS

Sebastian Riedel, <sri@cpan.org>

Christian Hansen, <ch@ngmedia.com>

Andy Grundman, <andy@hybridized.org>

=head1 COPYRIGHT

This program is free software, you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/Catalyst/Engine/Apache.pm
package Catalyst::Engine::Apache;

use strict;
use warnings;
use base 'Catalyst::Engine';

use File::Spec;
use URI;
use URI::http;
use URI::https;

use constant MP2 => ( 
    exists $ENV{MOD_PERL_API_VERSION} and 
           $ENV{MOD_PERL_API_VERSION} >= 2
);

our $VERSION = '1.15';

__PACKAGE__->mk_accessors(qw/apache return/);

sub prepare_request {
    my ( $self, $c, $r ) = @_;
    $self->apache( $r );
    $self->return( undef );
}

sub prepare_connection {
    my ( $self, $c ) = @_;

    $c->request->address( $self->apache->connection->remote_ip );

    PROXY_CHECK:
    {
        my $headers = $self->apache->headers_in;
        unless ( $c->config->{using_frontend_proxy} ) {
            last PROXY_CHECK if $c->request->address ne '127.0.0.1';
            last PROXY_CHECK if $c->config->{ignore_frontend_proxy};
        }        
        last PROXY_CHECK unless $headers->{'X-Forwarded-For'};

        # If we are running as a backend server, the user will always appear
        # as 127.0.0.1. Select the most recent upstream IP (last in the list)
        my ($ip) = $headers->{'X-Forwarded-For'} =~ /([^,\s]+)$/;
        $c->request->address( $ip );
    }

    $c->request->hostname( $self->apache->connection->remote_host );
    $c->request->protocol( $self->apache->protocol );
    $c->request->user( $self->apache->user );
    $c->request->remote_user( $self->apache->user );

    # when config options are set, check them here first
    if ($INC{'Apache2/ModSSL.pm'}) {
        $c->request->secure(1) if $self->apache->connection->is_https;
    } else {
        my $https = $self->apache->subprocess_env('HTTPS'); 
        $c->request->secure(1) if defined $https and uc $https eq 'ON';
    }

}

sub prepare_query_parameters {
    my ( $self, $c ) = @_;
    
    if ( my $query_string = $self->apache->args ) {
        $self->SUPER::prepare_query_parameters( $c, $query_string );
    }
}

sub prepare_headers {
    my ( $self, $c ) = @_;

    $c->request->method( $self->apache->method );

    if ( my %headers = %{ $self->apache->headers_in } ) {
        $c->request->header( %headers );
    }
}

sub prepare_path {
    my ( $self, $c ) = @_;

    my $scheme = $c->request->secure ? 'https' : 'http';
    my $host   = $self->apache->hostname || 'localhost';
    my $port   = $self->apache->get_server_port;

    # If we are running as a backend proxy, get the true hostname
    PROXY_CHECK:
    {
        unless ( $c->config->{using_frontend_proxy} ) {
            last PROXY_CHECK if $host !~ /localhost|127.0.0.1/;
            last PROXY_CHECK if $c->config->{ignore_frontend_proxy};
        }
        last PROXY_CHECK unless $c->request->header( 'X-Forwarded-Host' );
        
        $host = $c->request->header( 'X-Forwarded-Host' );

        if ( $host =~ /^(.+):(\d+)$/ ) {
            $host = $1;
            $port = $2;
        } else {
            # backend could be on any port, so
            # assume frontend is on the default port
            $port = $c->request->secure ? 443 : 80;
        }
    }

    my $base_path = '';

    # Are we running in a non-root Location block?
    my $location = $self->apache->location;
    if ( $location && $location ne '/' ) {
        $base_path = $location;
    }
    
    # Using URI directly is way too slow, so we construct the URLs manually
    my $uri_class = "URI::$scheme";
    
    if ( $port !~ /^(?:80|443)$/ && $host !~ /:/ ) {
        $host .= ":$port";
    }
    
    # We want the path before Apache escapes it.  Under mod_perl2 this is available
    # with the unparsed_uri method.  Under mod_perl 1 we must parse it out of the
    # request line.
    my ($path, $qs);
    
    if ( MP2 ) {
        ($path, $qs) = split /\?/, $self->apache->unparsed_uri, 2;
    }
    else {
        my (undef, $path_query) = split / /, $self->apache->the_request, 3;
        ($path, $qs)            = split /\?/, $path_query, 2;
    }
    
    # Don't check for LocationMatch blocks if requested
    # http://rt.cpan.org/Ticket/Display.html?id=26921
    if ( $self->apache->dir_config('CatalystDisableLocationMatch') ) {
        $base_path = '';
    }
        
    # Check if $base_path appears to be a regex (contains invalid characters),
    # meaning we're in a LocationMatch block
    elsif ( $base_path =~ m/[^$URI::uric]/o ) {
        # Find out what part of the URI path matches the LocationMatch regex,
        # that will become our base
        my $match = qr/($base_path)/;
        my ($base_match) = $path =~ $match;
        
        $base_path = $base_match || '';
    }

    # Strip leading slash
    $path =~ s{^/+}{};
    
    # base must end in a slash
    $base_path .= '/' unless $base_path =~ m{/$};

    # Are we an Apache::Registry script? Why anyone would ever want to run
    # this way is beyond me, but we'll support it!
    # XXX: This needs a test
    if ( defined $ENV{SCRIPT_NAME} && $self->apache->filename && -f $self->apache->filename && -x _ ) {
        $base_path .= $ENV{SCRIPT_NAME};
    }
    
    # If the path is contained within the base, we need to make the path
    # match base.  This handles the case where the app is running at /deep/path
    # but a request to /deep/path fails where /deep/path/ does not.
    if ( $base_path ne '/' && $base_path ne $path && $base_path =~ m{/$path} ) {
        $path = $base_path;
        $path =~ s{^/+}{};
    }
    
    my $query = $qs ? '?' . $qs : '';
    my $uri   = $scheme . '://' . $host . '/' . $path . $query;

    $c->request->uri( bless \$uri, $uri_class );
    
    my $base_uri = $scheme . '://' . $host . $base_path;

    $c->request->base( bless \$base_uri, $uri_class );
}

sub read_chunk {
    my $self = shift;
    my $c = shift;
    
    $self->apache->read( @_ );
}

sub finalize_body {
    my ( $self, $c ) = @_;
    
    $self->SUPER::finalize_body($c);
    
    # Data sent using $self->apache->print is buffered, so we need
    # to flush it after we are done writing.
    $self->apache->rflush;
}

sub finalize_headers {
    my ( $self, $c ) = @_;

    for my $name ( $c->response->headers->header_field_names ) {
        next if $name =~ /^Content-(Length|Type)$/i;
        my @values = $c->response->header($name);
        # allow X headers to persist on error
        if ( $name =~ /^X-/i ) {
            $self->apache->err_headers_out->add( $name => $_ ) for @values;
        }
        else {
            $self->apache->headers_out->add( $name => $_ ) for @values;
        }
    }

    # persist cookies on error responses
    if ( $c->response->header('Set-Cookie') && $c->response->status >= 400 ) {
        for my $cookie ( $c->response->header('Set-Cookie') ) {
            $self->apache->err_headers_out->add( 'Set-Cookie' => $cookie );
        }
    }

    # The trick with Apache is to set the status code in $apache->status but
    # always return the OK constant back to Apache from the handler.
    $self->apache->status( $c->response->status );
    $c->response->status( $self->return || $self->ok_constant );

    my $type = $c->response->header('Content-Type') || 'text/html';
    $self->apache->content_type( $type );

    if ( my $length = $c->response->content_length ) {
        $self->apache->set_content_length( $length );
    }

    return 0;
}

sub write {
    my ( $self, $c, $buffer ) = @_;

    if ( ! $self->apache->connection->aborted && defined $buffer) {
        return $self->apache->print( $buffer );
    }
    return;
}

1;
__END__

=head1 NAME

Catalyst::Engine::Apache - Catalyst Apache Engines

=head1 SYNOPSIS

For example Apache configurations, see the documentation for the engine that
corresponds to your Apache version.

C<Catalyst::Engine::Apache::MP13>  - mod_perl 1.3x

C<Catalyst::Engine::Apache2::MP19> - mod_perl 1.99x

C<Catalyst::Engine::Apache2::MP20> - mod_perl 2.x

=head1 DESCRIPTION

These classes provide mod_perl support for Catalyst.

=head1 METHODS

=head2 $c->engine->apache

Returns an C<Apache>, C<Apache::RequestRec> or C<Apache2::RequestRec> object,
depending on your mod_perl version.  This method is also available as
$c->apache.

=head2 $c->engine->return

If you need to return something other than OK from the mod_perl handler, 
you may set any other Apache constant in this method.  You should only use
this method if you know what you are doing or bad things may happen!
For example, to return DECLINED in mod_perl 2:

    use Apache2::Const -compile => qw(DECLINED);
    $c->engine->return( Apache2::Const::DECLINED );

=head2 NOTES ABOUT LOCATIONMATCH

The Apache engine tries to figure out the correct base path if your app is
running within a LocationMatch block.  For example:

    <LocationMatch ^/match/(this|that)*>
        SetHandler          modperl
        PerlResponseHandler MyApp
    </LocationMatch>

This will correctly set the base path to '/match/this/' or '/match/that/' depending
on which path was used for the request.

In some cases this may not be what you want, so you can disable this behavior
by adding this to your configuration:

    PerlSetVar CatalystDisableLocationMatch 1

=head2 NOTES ON NON-STANDARD PORTS

If you wish to run your site on a non-standard port you will need to use the
C<Port> Apache config rather than C<Listen>. This will result in the correct
port being added to urls created using C<uri_for>.

    Port 8080

=head1 OVERLOADED METHODS

This class overloads some methods from C<Catalyst::Engine>.

=over 4

=item prepare_request($r)

=item prepare_connection

=item prepare_query_parameters

=item prepare_headers

=item prepare_path

=item read_chunk

=item finalize_body

=item finalize_headers

=item write

=back

=head1 SEE ALSO

L<Catalyst> L<Catalyst::Engine>.

=head1 MAINTAINERS

Current maintainer, Tomas Doran (t0m) C<< <bobtfish@bobtfish.net> >>.

=head1 AUTHORS

Sebastian Riedel, C<< <sri@cpan.org> >>.

Christian Hansen, C<< <ch@ngmedia.com> >>.

Andy Grundman, C<< <andy@hybridized.org> >>.

=head1 COPYRIGHT

Copyright 2005-2010 by the above above listed "AUTHORS" and "MAINTAINERS".

=head1 LICENSE

This program is free software, you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
